File: leetcode_src/src/PathSum.java

public class PathSum {
    public boolean hasPathSum(TreeNode root, int targetSum) {
        if (root == null) return false;
        return helper(root, 0, targetSum);
    }

    private boolean helper(TreeNode node, int curr, int targetSum) {
        if (node == null) return false;
        else if(node.left == null && node.right == null) {
            return curr+node.val == targetSum;
        }
        else {
            if(helper(node.left, curr+node.val, targetSum) ||
             helper(node.right, curr+node.val, targetSum)) {
                return true;
            }
        }
        return false;
    }

    public static void main(String[] args) {
        TreeNode t1 = BinaryTreeUtil.generateBinaryTree("5,4,8,11,null,13,4,7,2,null,null,null,1");
        // TreeNode t1 = BinaryTreeUtil.generateBinaryTree("1,2");

        // t1.printByLayer();
        PathSum sol = new PathSum();
        System.out.println(sol.hasPathSum(t1, 22));
    }
}

File: leetcode_src/src/PathSum.py
from typing import *
from BinaryTreeUtil import *

class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        stack = [(root, 0)]
        if not root: return False

        run = 0
        while stack:
            node, run = stack.pop()
            if node.left:
                stack.append((node.left, run+node.val))
            if node.right:
                stack.append((node.right, run+node.val))

            if not node.left and not node.right and run + node.val == targetSum:
                return True

        return False

if __name__ == '__main__':
    t1 = generateBinaryTree("5,4,8,11,null,13,4,7,2,null,null,null, null,null,1")
    print(inorder(t1))

    print(Solution().hasPathSum(t1, 18))
